capsules/src/radio.rs

```diff
@@ -144,42 +139,66 @@ impl<'a, R: radio::Radio> Driver for RadioDriver<'a, R> {
                         rval = ReturnCode::ESIZE;
                         return;
                     }
-                    let offset = self.radio.payload_offset(false, false) as usize;
-                    // Copy the packet into the kernel buffer
-                    self.kernel_tx.map(|kbuf| {
-                        app.app_write.as_mut().map(|src| {
-                            for (i, c) in src.as_ref()[0..len].iter().enumerate() {
-                                kbuf[i + offset] = *c;
+
+                    // Copy the packet into the kernel frame
+                    let frame_info = self.kernel_tx.map_or(None, |tx_buf| {
+                        // Prepare frame headers
+                        let pan = self.mac.get_pan();
+                        let src_addr = MacAddress::Short(self.mac.get_address());
+                        let mut frame_info = match self.mac.prepare_data_frame(
+                            tx_buf, pan, MacAddress::Short(addr), pan, src_addr,
+                            None) {
+                            Ok(info) => info,
+                            Err(_) => {
+                                rval = ReturnCode::FAIL;
+                                return None;
                             }
+                        };
+
+                        // Copy the payload from userspace into kernelspace
+                        app.app_write.as_mut().map(|src| {
+                            rval = frame_info.append_payload(tx_buf,
+                                                             &src.as_ref()[..len]);
                         });
+                        Some(frame_info)
                     });
-                    let kbuf = self.kernel_tx.take().unwrap();
-                    rval = self.radio.transmit(addr, kbuf, len as u8, false);
-                    if rval == ReturnCode::SUCCESS {
-                        self.busy.set(true);
+                    if rval != ReturnCode::SUCCESS {
+                        return;
+                    }
+
+                    // Try to transmit the frame, otherwise at least get the
+                    // frame back.
+                    let res = self.mac.transmit(self.kernel_tx.take().unwrap(),
```

Comment:
`unwrap` is a no-no. I think it can be totally avoided by putting these calls in the close above that, anyway, maps over `kernel_tx`.